src/ray/ray_impl.rs

```diff
@@ -113,31 +113,26 @@ impl<T: BHValue, const D: usize> Ray<T, D> {
     /// where the first number is the distance from [`Ray`] to the nearest intersection point
     /// and the second number is the distance from [`Ray`] to the farthest intersection point
     ///
-    /// If there are no intersections, it returns negative one for both distances
-    pub fn intersection_slice_for_aabb(&self, aabb: &Aabb<T, D>) -> (T, T)
+    /// If there are no intersections, it returns `None`.
+    pub fn intersection_slice_for_aabb(&self, aabb: &Aabb<T, D>) -> Option<(T, T)>
     where
         T: BHValue,
     {
-        // https://iquilezles.org/articles/intersectors/
-        let mut n = self.origin.coords - aabb.center().coords;
-        n.component_mul_assign(&self.inv_direction);
+        // Copied from the default `RayIntersection` implementation. TODO: abstract and add SIMD.
+        let lbr = (aabb[0].coords - self.origin.coords).component_mul(&self.inv_direction);
+        let rtr = (aabb[1].coords - self.origin.coords).component_mul(&self.inv_direction);
 
-        let k = self.inv_direction.abs().component_mul(&aabb.half_size());
-        let t1 = -n - k;
-        let t2 = -n + k;
+        let (inf, sup) = lbr.inf_sup(&rtr);
 
-        let entry_distance = t1
-            .iter()
-            .skip(1)
-            .fold(t1[0], |acc, x| -> T { fast_max(*x, acc) });
-        let exit_distance = t2.iter().skip(1).fold(t2[0], |acc, x| fast_min(*x, acc));
+        let tmin = inf.max();
+        let tmax = sup.min();
 
         // no intersection
-        if entry_distance > exit_distance || exit_distance < T::zero() {
-            return (T::from_f32(-1.0).unwrap(), T::from_f32(-1.0).unwrap());
+        if tmin > tmax || tmax < T::zero() {
+            return None;
         }
 
-        (fast_max(entry_distance, T::zero()), exit_distance)
+        Some((fast_max(tmin, T::zero()), tmax))
     }
```

Comment:
I'll be honest, I didn't check the math but I trust you since the other tests still work. :P

Reply:
Don't need to trust the math, since I added a test that the distances match those measured in blender for entry and exit. They were an exact match, modulo precision issues. As the comment states, the math is the same as ray-AABB intersection, modified to return the distance rather than comparing it.